Update Everest updater to align with build pipeline rework

coloursofnoise opened the pull request:
Each branch is now a different source, and is displayed separately in the version list.
This allows for more flexibility in what branches are included, opening the possibility of adding feature branches in the future.

Stable and beta builds are taken from GitHub Releases to align with the changes in #498.

I believe this can be safely merged before merging the pipeline changes, although it might be a good idea to manually create a few beta releases on GitHub to ease the transition.

This PR should also be tested on Windows (and probably Mac just in case) to make sure the in-game Updater still functions as intended.

coloursofnoise committed:
Add URIHelper for making query string from NameValuePair

Could be replaced with multi-line string interpolation in C#11

coloursofnoise committed:
Add TextMenu BatchMode context and SubHeaderExt AlwaysCenter

coloursofnoise committed:
Redo Everest updater and version selection to separate branches

Each branch is a different source, allowing beta and stable
releases published on GitHub, and better support for adding
feature branches in future.

coloursofnoise committed:
Fix TitleScreen update notification

coloursofnoise committed:
Add Dialog keys for new updater text

coloursofnoise committed:
Format changed files

coloursofnoise committed:
Add User-Agent header to all web requests

coloursofnoise committed:
Use User-Agent property to set header

coloursofnoise committed:
Set Accept header to tell GitHub to download asset binary

coloursofnoise committed:
Make Everest.Updater class accessible to mods

coloursofnoise committed:
Rename fields for clarity and fix display

DemoJameson commented on Celeste.Mod.mm/Mod/Everest/Everest.Updater.cs:
```diff
@@ -232,27 +229,47 @@ public class Source {
+                        string url = null;
+                        foreach (JObject asset in release["assets"] as JArray) {
+                            if (asset["name"].ToString() == "main.zip") {
+                                url = asset["url"].ToString();
```
should use `browser_download_url` instead of `url`

coloursofnoise replied:
This has been changed, although the GitHub Release Parser is no longer used for the main Everest branches.

DemoJameson commented on Celeste.Mod.mm/Mod/Everest/Everest.Updater.cs:
```diff
@@ -159,14 +136,34 @@ public class Source {
 
-                    ParseData = AzureDataParser("https://dev.azure.com/EverestAPI/Everest/_apis/build/builds/{0}/artifacts?artifactName=main&api-version=5.0&%24format=zip", 700)
-                }
+                    Index = "https://api.github.com/repos/EverestAPI/Everest/releases",
```
Is there a way to avoid triggering the API request cap?
Get this error when I try accessing this url via browser.

```json
{"message":"API rate limit exceeded for 104.134.74.233. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
```

WEGFan replied:
> For unauthenticated requests, the rate limit allows for up to 60 requests per hour. Unauthenticated requests are associated with the originating IP address, and not the person making requests.

🤔 Maybe we can ask max4805 to fetch releases with access token and cache the results?

maddie480 replied:
I think API rate limits is also why Olympus shows a limited amount of commit descriptions in the versions list 🤔 I could definitely fetch and cache the list on my end, the only issue would be that there would be a small delay before a new release appears

WEGFan replied:
> There would be a small delay before a new release appears.

We can use webhooks, also the limit is 5000 requests per hour with access tokens so I think we can safely increase fetching frequency.

maddie480 replied:
We can either have an API that gives [a preformatted list of versions](https://cdn.discordapp.com/attachments/429775439423209472/1010647544541499512/a.json) that could be used by Everest and Olympus to avoid maintaining the logic twice, or just have an API that returns the releases as-is to keep all logic client-side, what do we prefer? 🤔

maddie480 replied:
https://max480-random-stuff.appspot.com/celeste/everest-versions ^^

coloursofnoise replied:
Has been implemented using https://everestapi.github.io/everestupdater.txt to link to the update list source

coloursofnoise committed:
Use browser_download_url when downloading builds from GitHub

coloursofnoise committed:
Use pre-formatted everest update list

Azure and GitHub parsers are kept for use by mods

coloursofnoise committed:
Increase max number of builds on update screen

Each of the three branches have been moved to their own tab,
so a hard cap is much less useful.

DemoJameson commented on Celeste.Mod.mm/Mod/Everest/Everest.Updater.cs:
```diff
@@ -277,6 +314,7 @@ public class Source {
                     return;
                 }
 
+                CoreModule.Settings.CurrentBranch = version.Source.Name;
```
This setting is not actually written to the file
perhaps move it until after the installation succeeds and call `CoreModule.Instance.SaveSettings()` manually

coloursofnoise committed:
Explicitly save CurrentBranch setting

coloursofnoise committed:
Merge dev branch into pipeline-rework-updater in preparation for merge
